membership-attribute-service/app/controllers/SalesforceHookController.scala

```diff
@@ -39,48 +39,65 @@ class SalesforceHookController extends LazyLogging {
-          if (tierFromZuora != tierFromSalesforceWebhook) logger.error(s"Differing tier info for $sfId : sf=$tierFromSalesforceWebhook zuora=$tierFromZuora")
-          attrs.copy(Tier = tierFromZuora)
-        }).run.flatMap { attrsUpdatedWithZuoraOpt =>
-          if (attrsUpdatedWithZuoraOpt.isEmpty) logger.error(s"Couldn't update $attrs with information from Zuora")
-          attributeService.set(attrsUpdatedWithZuoraOpt.getOrElse(attrs)).map { putItemResult =>
-            logger.info(s"Successfully inserted ${attrsUpdatedWithZuoraOpt.getOrElse(attrs)} into ${touchpoint.dynamoTable}.")
-            metrics.put("Update", 1)
-            ack
-          }.recover { case e: Throwable =>
-            logger.warn(s"Failed to insert ${attrsUpdatedWithZuoraOpt.getOrElse(attrs)} into ${touchpoint.dynamoTable}. Salesforce should retry.", e)
-            ApiErrors.internalError
-          }
-        }
-      case -\/(ParsingError(msg)) =>
-        Logger.error(s"Could not parse payload ${request.body}:\n$msg")
-        Future(ApiErrors.badRequest(msg))
-      case -\/(OrgIdMatchingError(orgId)) =>
-        Logger.error(s"Wrong organization Id: $orgId")
-        Future(ApiErrors.unauthorized.copy("Wrong organization Id"))
+  val touchpoint = request.touchpoint
+  val validOrgId = touchpoint.sfOrganisationId
+  val attributeService = touchpoint.attrService
+  implicit val pf = Membership
+
+  def processDeletion(membershipDeletion: MembershipDeletion) = {
+    val userId = membershipDeletion.userId
+    attributeService.delete(userId).map { x =>
+      logger.info(s"Successfully deleted user $userId from ${touchpoint.dynamoTable}.")
+      metrics.put("Delete", 1)
+      Success
+    }.recover { case e: Throwable =>
+      logger.warn(s"Failed to delete user $userId from ${touchpoint.dynamoTable}. Salesforce should retry.", e)
+      Failure
+    }
+  }
+
+  def processUpdate(membershipUpdate: MembershipUpdate) = {
+    val attrs = membershipUpdate.attributes
+    (for {
+      sfId <- OptionT(touchpoint.contactRepo.get(attrs.UserId))
+      membershipSubscription <- OptionT(touchpoint.subService.current[SubscriptionPlan.Member](sfId).map(_.headOption))
+    } yield {
+      val tierFromSalesforceWebhook = attrs.Tier
+      val tierFromZuora = membershipSubscription.plan.charges.benefit.id
+      if (tierFromZuora != tierFromSalesforceWebhook) logger.error(s"Differing tier info for $sfId : sf=$tierFromSalesforceWebhook zuora=$tierFromZuora")
+      attrs.copy(Tier = tierFromZuora)
+    }).run.flatMap { attrsUpdatedWithZuoraOpt =>
+      if (attrsUpdatedWithZuoraOpt.isEmpty) logger.error(s"Couldn't update $attrs with information from Zuora")
+      attributeService.set(attrsUpdatedWithZuoraOpt.getOrElse(attrs)).map { putItemResult =>
+        logger.info(s"Successfully inserted ${attrsUpdatedWithZuoraOpt.getOrElse(attrs)} into ${touchpoint.dynamoTable}.")
+        metrics.put("Update", 1)
+        Success
+      }.recover { case e: Throwable =>
```

Comment:
in an ideal world we would wrap anything that might throw an exception with a Try and turn the result into a `\/` straight away, to avoid having exception handling code dotted around